modules/core/src/shadertools/src/modules/phong-lighting/phong-lighting.js

```diff
@@ -0,0 +1,96 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import LightingEffect from '../../../../lighting/lighting-effect';
+import lightingShader from './phong-lighting.glsl';
+
+export default {
+  name: 'lighting',
+  vs: lightingShader,
+  getUniforms
+};
+
+const INITIAL_MODULE_OPTIONS = {};
+
+function getLightSourceUniforms(lightEffect) {
+  const lightSourceUniforms = {};
+
+  if (lightEffect.ambientLight) {
+    lightSourceUniforms['lighting_ambientLight.color'] = lightEffect.ambientLight.color;
+    lightSourceUniforms['lighting_ambientLight.intensity'] = lightEffect.ambientLight.intensity;
+  }
+
+  let index = 0;
+  for (const i in lightEffect.pointLights) {
+    const pointLight = lightEffect.pointLights[i];
+    lightSourceUniforms[`lighting_pointLight[${index}].color`] = pointLight.color;
+    lightSourceUniforms[`lighting_pointLight[${index}].intensity`] = pointLight.intensity;
+    lightSourceUniforms[`lighting_pointLight[${index}].position`] = pointLight.position;
+    index++;
+  }
+  lightSourceUniforms.lighting_pointLightNumber = index;
+
+  for (const i in lightEffect.directionalLights) {
+    const directionalLight = lightEffect.directionalLights[i];
+    lightSourceUniforms[`lighting_directionalLight[${index}].color`] = directionalLight.color;
+    lightSourceUniforms[`lighting_directionalLight[${index}].intensity`] =
+      directionalLight.intensity;
+    lightSourceUniforms[`lighting_directionalLight[${index}].direction`] =
+      directionalLight.direction;
+    index++;
+  }
+  lightSourceUniforms.lighting_directionalLightNumber = lightEffect.directionalLights.length;
+
+  return lightSourceUniforms;
+}
+
+function getMaterialUniforms(material) {
+  const materialUniforms = {};
+  materialUniforms.lighting_ambient = material.ambient;
+  materialUniforms.lighting_diffuse = material.diffuse;
+  materialUniforms.lighting_shininess = material.shininess;
+  materialUniforms.lighting_specularColor = material.specularColor;
+  return materialUniforms;
+}
+
+function getUniforms(opts = INITIAL_MODULE_OPTIONS) {
+  let lightEffect;
+  if (opts.effects && Array.isArray(opts.effects)) {
+    for (const i in opts.effects) {
+      const effect = opts.effects[i];
+      if (effect instanceof LightingEffect) {
+        lightEffect = effect;
+        break;
+      }
+    }
+  }
+
+  if (!lightEffect || !opts.material) {
+    return {};
+  }
+
+  const lightUniforms = Object.assign(
+    {},
+    getLightSourceUniforms(lightEffect),
```

Comment:
Is this going to be called for every model that uses light effect? Cache the result so we don't do it over and over agin?